### ISSUE TITLE
OpenCV disabling libopenh264 for ffmpeg

### ISSUE THREAD

**Issue by fvivian** [conda-forge/opencv-feedstock#131]
Issue:
Whenever I install opencv via conda-forge, it disables libx264 and libopenh264. I try to create mp4 videos with opencv's VideoWriter to send it via Telegram Chat Bot. Telegram requires the media file to be encoded with H.264 and the container should be mp4. I tried different codecs/containers but the file could not be viewed on the phone, therefore, I try to make it run with these prerequisites.

The error message when using <code>fourcc = cv2.VideoWriter_fourcc(*'H264')</code>: 
<code>
OpenCV: FFMPEG: tag 0x34363248/'H264' is not supported with codec id 27 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x31637661/'avc1'
Could not find encoder for codec id 27: Encoder not found
</code>

After some research I found out that libopenh264 for ffmpeg has to be enabled in order to use the h264 codec. To check this I ran <code>ffmpeg</code> which returns;
<code>
ffmpeg version 4.0 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 7.2.0 (crosstool-NG fa8859cb)
  configuration: --prefix=/home/eouser/anaconda3/envs/gif --cc=/opt/conda/conda-bld/ffmpeg_1531088893642/_build_env/bin/x86_64-conda_cos6-linux-gnu-cc --disable-doc --enable-shared --enable-static --enable-zlib --enable-pic --enable-gpl --enable-version3 --disable-nonfree --enable-hardcoded-tables --enable-avresample --enable-libfreetype --disable-openssl --disable-gnutls --enable-libvpx --enable-pthreads --enable-libopus --enable-postproc --disable-libx264
</code>

I then uninstalled ffmpeg and opencv with it. Reinstalling ffmpeg alone via cond-forge and checking the flags with <code>ffmpeg</code> again returns;
<code>
ffmpeg version 4.0.2 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 4.8.2 (GCC) 20140120 (Red Hat 4.8.2-15)
  configuration: --prefix=/home/eouser/anaconda3/envs/gif --disable-doc --disable-openssl --enable-shared --enable-static --extra-cflags='-Wall -g -m64 -pipe -O3 -march=x86-64 -fPIC' --extra-cxxflags='-Wall -g -m64 -pipe -O3 -march=x86-64 -fPIC' --extra-libs='-lpthread -lm -lz' --enable-zlib --enable-pic --enable-pthreads --enable-gpl --enable-version3 --enable-hardcoded-tables --enable-avresample --enable-libfreetype --enable-gnutls --enable-libx264 --enable-libopenh264
</code>

which looks ok. After installing opencv via conda-forge I get the following result for ffmpeg:
<code>
ffmpeg version 4.0 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 7.2.0 (crosstool-NG fa8859cb)
  configuration: --prefix=/home/eouser/anaconda3/envs/gif --cc=/opt/conda/conda-bld/ffmpeg_1531088893642/_build_env/bin/x86_64-conda_cos6-linux-gnu-cc --disable-doc --enable-shared --enable-static --enable-zlib --enable-pic --enable-gpl --enable-version3 --disable-nonfree --enable-hardcoded-tables --enable-avresample --enable-libfreetype --disable-openssl --disable-gnutls --enable-libvpx --enable-pthreads --enable-libopus --enable-postproc --disable-libx264
</code>

How do I enable openh264 after having installed opencv?

<code>

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /home/eouser/anaconda3/envs/gif:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py37_0
blas                      1.0                         mkl
bzip2                     1.0.6                h14c3975_5
ca-certificates           2018.8.24            ha4d7672_0    conda-forge
cairo                     1.14.12              h8948797_3
certifi                   2018.8.24             py37_1001    conda-forge
cffi                      1.11.5           py37he75722e_1
chardet                   3.0.4                    py37_1
cryptography              2.3.1            py37hc365091_0
ffmpeg                    4.0                  hcdf2ecd_0
fontconfig                2.13.0               h9420a91_0
freeglut                  3.0.0                hf484d3e_5
freetype                  2.9.1                h6debe1e_4    conda-forge
glib                      2.56.2               hd408876_0
gmp                       6.1.2                hfc679d8_0    conda-forge
gnutls                    3.5.19               h2a4e5f8_1    conda-forge
graphite2                 1.3.12               h23475e2_2
harfbuzz                  1.9.0                hee26f79_1    conda-forge
hdf5                      1.10.2               hba1933b_1
icu                       58.2                 h9c2bf20_1
idna                      2.7                      py37_0
intel-openmp              2018.0.3                      0
jasper                    2.0.14               h07fcdf6_1
jpeg                      9c                   h470a237_1    conda-forge
libedit                   3.1.20170329         h6b74fdf_2
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 8.2.0                hdf63c60_1
libgfortran-ng            7.3.0                hdf63c60_0
libglu                    9.0.0                hf484d3e_1
libiconv                  1.15                 h470a237_3    conda-forge
libopencv                 3.4.2                hb342d67_1
libopus                   1.2.1                hb9ed12e_0
libpng                    1.6.34               hb9fc6fc_0
libprotobuf               3.5.2                hd28b015_1    conda-forge
libstdcxx-ng              8.2.0                hdf63c60_1
libtiff                   4.0.9                he85c1e1_2
libuuid                   1.0.3                h1bed415_2
libvpx                    1.7.0                h439df22_0
libxcb                    1.13                 h1bed415_1
libxml2                   2.9.8                h26e45fe_1
mkl                       2018.0.3                      1
mkl_fft                   1.0.4            py37h4414c95_1
mkl_random                1.0.1            py37h4414c95_1
ncurses                   6.1                  hf484d3e_0
nettle                    3.3                           0    conda-forge
numpy                     1.15.1           py37h1d66e8a_0
numpy-base                1.15.1           py37h81de0dd_0
olefile                   0.45.1                     py_1    conda-forge
opencv                    3.4.2            py37h6fd60c2_1
openh264                  1.7.0                         0    conda-forge
openssl                   1.0.2p               h470a237_0    conda-forge
pcre                      8.42                 h439df22_0
pillow                    5.3.0            py37hc736899_0    conda-forge
pip                       10.0.1                   py37_0
pixman                    0.34.0               hceecf20_3
proj4                     5.0.1                h14c3975_0
py-opencv                 3.4.2            py37hb342d67_1
pycparser                 2.18                     py37_1
pyopenssl                 18.0.0                   py37_0
pyproj                    1.9.5.1          py37h7b21b82_1
pysocks                   1.6.8                    py37_0
python                    3.7.0                hc3d631a_0
readline                  7.0                  h7b6447c_5
requests                  2.19.1                   py37_0
setuptools                40.2.0                   py37_0
six                       1.11.0                   py37_1
sqlite                    3.24.0               h84994c4_0
tk                        8.6.8                hbc83047_0
urllib3                   1.23                     py37_0
wheel                     0.31.1                   py37_0
x264                      1!152.20180717       h470a237_1    conda-forge
xz                        5.2.4                h14c3975_4
zlib                      1.2.11               ha838bed_2


```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : gif
    active env location : /home/eouser/anaconda3/envs/gif
            shell level : 2
       user config file : /home/eouser/.condarc
 populated config files :
          conda version : 4.5.11
    conda-build version : 3.4.1
         python version : 3.6.4.final.0
       base environment : /home/eouser/anaconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /home/eouser/anaconda3/pkgs
                          /home/eouser/.conda/pkgs
       envs directories : /home/eouser/anaconda3/envs
                          /home/eouser/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.11 requests/2.18.4 CPython/3.6.4 Linux/4.4.0-96-generic ubuntu/16.04 glibc/2.23
                UID:GID : 1001:1001
             netrc file : None
           offline mode : False

```
</details>

**Comment by hmaarrfk** [comment]
@fvivian thank you sharing your issue. 

I'm not an expert in encoding. Could you please provide us with a short example that we can run on our computers to reproduce the problem?

Generally speaking, you should avoid mixing `defaults` and `conda-forge` as it produces certain incompatibilities. For pure python this isn't really an issue, but for things like this (opencv is largely C), it might be due to that.

I would create a fresh environment and only install conda-forge things.

If you have trouble with numpy or scipy being chosen from `defaults`:

```
conda install conda-forge::numpy
```
should do the trick 

Please also note that conda-forge still doesn't do well with python3.7. You should create a python 3.6 environment (python 3.5 support seems to have been dropped yesterday!)

**Comment by fvivian** [comment]
@hmaarrfk thanks for your quick reply. This is the script I use to create the video: https://github.com/fvivian/TerraMaterBot/s5p/createVid.py but I'm afraid you probably cannot use it on your computer, so I will try to draft a short example for you to run. For this I will also create a new environment as you suggested and let you know how that worked out.

**Comment by fvivian** [comment]
Alright, I followed your advice and created an environment with python 3.6 and installed all the packages via conda-forge. The great news is, the video.mp4 is created and looks fine in VLC. The only issue I see now is the fallback by ffmpeg when running the script:
<code>
OpenCV: FFMPEG: tag 0x34363248/'H264' is not supported with codec id 27 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x31637661/'avc1'
</code>

I will have to test this in Telegram and see if the video can be opened on a smartphone.

tl;dr: The question "How do I enable openh264 after having installed opencv?" can be answered by using python 3.6 and install all the packages via conda-forge. If the video is created with the codec H264 remains to be seen, at least the container is mp4.  I guess this issue can be closed.

**Comment by hmaarrfk** [comment]
It is very possible that h264 support hasn't fully been built in.
I don't think adding to conda-forge is out of scope. 
But help in adding it would be appreciated.

**Comment by suhrmann** [comment]
I fixed this issue by replacing `ffmpeg` of conda with built in of Ubuntu 20.04

``conda remove ffmpeg --force `` broke `opencv` (missing / broken library)

     
> `--force-remove`, `--force`
> Forces removal of a package without removing packages that depend on it. Using this option will usually leave your environment in a broken and inconsistent state.

Source: [Conda command referene >> `conda remove`](https://docs.conda.io/projects/conda/en/latest/commands/remove.html)


## Setup Instructions

### Environment and package

Create Python 3.7 env using conda and fix / replace ``ffmpeg``

Create env and install OpenCV, `ffmpeg` (without h264) is a dependency of it
```bash
# Create environment "OpenCV"
conda create -n OpenCV python=3.7
conda activate OpenCV

# Install OpenCV using `conda`
conda install opencv  
```

Running ``ffmpeg`` prints configuration flag ``--disable-libx264``:
<details>
<summary><code>$ ffmpeg</code> of Conda</summary>

<p>

```bash

$ ffmpeg
ffmpeg version 4.0 Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 7.2.0 (crosstool-NG fa8859cb)
  configuration: --prefix=/home/simon/.conda/envs/OpenCV --cc=/opt/conda/conda-bld/ffmpeg_1531088893642/_build_env/bin/x86_64-conda_cos6-linux-gnu-cc --disable-doc --enable-shared --enable-static --enable-zlib --enable-pic --enable-gpl --enable-version3 --disable-nonfree --enable-hardcoded-tables --enable-avresample --enable-libfreetype --disable-openssl --disable-gnutls --enable-libvpx --enable-pthreads --enable-libopus --enable-postproc --disable-libx264
  libavutil      56. 14.100 / 56. 14.100
  libavcodec     58. 18.100 / 58. 18.100
  libavformat    58. 12.100 / 58. 12.100
  libavdevice    58.  3.100 / 58.  3.100
  libavfilter     7. 16.100 /  7. 16.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  1.100 /  5.  1.100
  libswresample   3.  1.100 /  3.  1.100
  libpostproc    55.  1.100 / 55.  1.100
Hyper fast Audio and Video encoder
usage: ffmpeg [options] [[infile options] -i infile]... {[outfile options] outfile}...

Use -h to get full help or, even better, run 'man ffmpeg'
```

**Note** 

</p>
</details>


**Package list of env ``OpenCV``:**
<details>
<summary><code>(OpenCV)$ conda list</code></summary>

<p>

```bash
# packages in environment at /home/simon/.conda/envs/OpenCV:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
blas                      1.0                         mkl  
bzip2                     1.0.8                h7b6447c_0  
ca-certificates           2020.1.1                      0  
cairo                     1.14.12              h8948797_3  
certifi                   2020.4.5.1               py37_0  
ffmpeg                    4.0                  hcdf2ecd_0  
fontconfig                2.13.0               h9420a91_0  
freeglut                  3.0.0                hf484d3e_5  
freetype                  2.9.1                h8a8886c_1  
glib                      2.63.1               h5a9c865_0  
graphite2                 1.3.13               h23475e2_0  
harfbuzz                  1.8.8                hffaf4a1_0  
hdf5                      1.10.2               hba1933b_1  
icu                       58.2                 he6710b0_3  
intel-openmp              2020.0                      166  
jasper                    2.0.14               h07fcdf6_1  
jpeg                      9b                   h024ee3a_2  
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20181209         hc058e9b_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libglu                    9.0.0                hf484d3e_1  
libopencv                 3.4.2                hb342d67_1  
libopus                   1.3.1                h7b6447c_0  
libpng                    1.6.37               hbc83047_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.1.0                h2733197_0  
libuuid                   1.0.3                h1bed415_2  
libvpx                    1.7.0                h439df22_0  
libxcb                    1.13                 h1bed415_1  
libxml2                   2.9.9                hea5a465_1  
mkl                       2020.0                      166  
mkl-service               2.3.0            py37he904b0f_0  
mkl_fft                   1.0.15           py37ha843d7b_0  
mkl_random                1.1.0            py37hd6b4f25_0  
ncurses                   6.2                  he6710b0_1  
numpy                     1.18.1           py37h4f9e942_0  
numpy-base                1.18.1           py37hde5b4d6_1  
opencv                    3.4.2            py37h6fd60c2_1  
openssl                   1.1.1g               h7b6447c_0  
pcre                      8.43                 he6710b0_0  
pip                       20.0.2                   py37_1  
pixman                    0.38.0               h7b6447c_0  
py-opencv                 3.4.2            py37hb342d67_1  
python                    3.7.7           hcf32534_0_cpython  
readline                  8.0                  h7b6447c_0  
setuptools                46.1.3                   py37_0  
six                       1.14.0                   py37_0  
sqlite                    3.31.1               h62c20be_1  
tk                        8.6.8                hbc83047_0  
wheel                     0.34.2                   py37_0  
xz                        5.2.5                h7b6447c_0  
zlib                      1.2.11               h7b6447c_3  
zstd                      1.3.7                h0b5b093_0  
```
</p>
</details>  




## Fix / workaround `ffmpeg`


### Install ``ffmpeg`` with ``apt``
Install ``ffmpeg`` with ``apt`` - and I also installed ``ubuntu-restricted-extras`` to be safe:
```bash
sudo apt-get install ffmpeg ubuntu-restricted-extras
```

### Replace conda ``ffmpeg`` with built-in ``ffmpeg`` of Ubuntu

```bash
# "disable" conda ffmpeg 
which ffmpeg  # -> Output: /home/simon/.conda/envs/OpenCV/bin/ffmpeg
cd /home/simon/.conda/envs/OpenCV/bin
mv ffmpeg ffmpeg_BAK
# Replace with Ubuntu ffmpeg
which ffmpeg  # Expecting path to Ubuntu ffmpeg - Output: /usr/bin/ffmpeg
ln -s /usr/bin/ffmpeg ffmpeg  # Link conda ffmpeg to built in
```


### Results

**Output of ``ffmpeg`` - in activated env ``OpenCV``:**
<details>
<summary><b>Expected</b> output of <code>which ffmpeg</code> and <code>ffmpeg</code></summary>

<p>

```bash
$ which ffmpeg
/home/simon/.conda/envs/OpenCV/bin/ffmpeg

$ ffmpeg
ffmpeg version 4.2.2-1ubuntu1 Copyright (c) 2000-2019 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.3.0-3ubuntu1)
  configuration: --prefix=/usr --extra-version=1ubuntu1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample --disable-filter=resample --enable-avisynth --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
  libavutil      56. 31.100 / 56. 31.100
  libavcodec     58. 54.100 / 58. 54.100
  libavformat    58. 29.100 / 58. 29.100
  libavdevice    58.  8.100 / 58.  8.100
  libavfilter     7. 57.100 /  7. 57.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  5.100 /  5.  5.100
  libswresample   3.  5.100 /  3.  5.100
  libpostproc    55.  5.100 / 55.  5.100
Hyper fast Audio and Video encoder
usage: ffmpeg [options] [[infile options] -i infile]... {[outfile options] outfile}...

Use -h to get full help or, even better, run 'man ffmpeg'
```

**Note** 
 - We successfully replaced conda `ffmpeg` with Ubuntu `ffmpeg` due to path of Ubuntu and version change (OpenCV was shipped with `ffmpeg=4.0`)
 - Search for configuration ``--enable-libx264``, so h264 is working now - finally -_-

</p>
</details>

**Comment by hmaarrfk** [comment]
It seems that you are finding something interesting and that it is likely that we should revive this issue, in a new, separate issue

**Comment by suhrmann** [comment]
Is this a polite demand for me to create this issue or were you just thinking aloud?

**Comment by hmaarrfk** [comment]
Yes. It is a polite way for me to ask.

I can create it myself, but typically issues from 2018 are no longer the same issues we are dealing with today on conda forge.

You also seem to have found a test case that proves that the functionality you need isn't included.

Since this issue is currently closed, your challenges will be ignore if I forget to check back due to other priorities taking my time (and that of other contributors)

**Comment by hmaarrfk** [comment]
If you create it yourself, you will also automatically get pinged on progress.
